File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        priority_queue<pair<int, int>> pq;
        unordered_map<int, int> counter;
        vector<int> result;

        for (auto& num: nums) {
            counter[num] += 1;
        }

        for (auto& pair: counter) {
            int value = pair.first;
            int freq = pair.second;
            pq.push({freq, value});
        }

        while (!pq.empty() && k>0) {
            int value = pq.top().second;
            result.push_back(value);
            pq.pop();
            k -= 1;
        }

        return result;
        
    }
};

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        '''
        first count freq of nums in a hashmap
        put values of map into heap of size k

        '''

        min_heap = []
        freq_map = {}
        res = []

        for num in nums:
            freq_map[num] = freq_map.get(num, 0) + 1

        for key, value in freq_map.items():
            heapq.heappush(min_heap, (value, key))

            if len(min_heap) > k:
                heapq.heappop(min_heap)
        
        while min_heap:
            res.append(heapq.heappop(min_heap)[1])

        return res